components/raftstore/src/store/fsm/apply.rs

```diff
@@ -2658,6 +2702,53 @@ impl ApplyFsm {
         );
     }
 
+    fn handle_change(&mut self, apply_ctx: &mut ApplyContext, change: ChangeCmd) {
+        let (region_id, region_epoch, cb) = match change {
+            ChangeCmd::RegisterObserver {
+                region_id,
+                region_epoch,
+                enabled,
+                cb,
+            } => {
+                assert!(!self
+                    .delegate
+                    .observe_cmd
+                    .as_ref()
+                    .map_or(false, |e| e.load(Ordering::Relaxed)));
+                // TODO(cdc): take cmd_observer_enabled when enabled is false.
+                self.delegate.observe_cmd = Some(enabled);
+                (region_id, region_epoch, cb)
+            }
+            ChangeCmd::Snapshot {
+                region_id,
+                region_epoch,
+                cb,
+            } => (region_id, region_epoch, cb),
+        };
+
+        assert_eq!(self.delegate.region_id(), region_id);
+        let resp = match compare_region_epoch(
+            &region_epoch,
+            &self.delegate.region,
+            false, /* check_conf_ver */
+            true,  /* check_ver */
+            true,  /* include_region */
+        ) {
+            Ok(()) => ReadResponse {
+                response: Default::default(),
+                snapshot: Some(RegionSnapshot::<RocksEngine>::from_raw(
+                    apply_ctx.engines.kv.clone(),
+                    self.delegate.region.clone(),
+                )),
+            },
+            Err(e) => ReadResponse {
+                response: cmd_resp::new_error(e),
+                snapshot: None,
+            },
+        };
+        cb.invoke_read(resp);
```

Comment:
Should we flush write batch before invoking read?

Reply:
Actually we should flush write batch before getting snapshot.